.vsts-pipelines/templates/phases/default-build.yml

```diff
@@ -0,0 +1,43 @@
+# default-build.yml
+# Description: Defines a build phase for invoking build.sh/cmd
+# Parameters:
+#   agentOs: [Windows_NT (default), Linux, macOS]
+#       Used in templates to define variables which are OS specific
+#   beforeBuild: [steps]
+#       Additional steps to run before build.sh/cmd
+#   afterBuild: [steps]
+#       Additional steps to run after build.sh/cmd
+
+parameters:
+  agentOs: ''
+  buildArgs: ''
+  beforeBuild: []
+  afterBuild: []
+
+phases:
+- phase: ${{ parameters.os }}
+  queue:
+    ${{ if eq(parameters.os, 'macOS') }}:
+      name: Hosted macOS Preview
+    ${{ if eq(parameters.os, 'Linux') }}:
+      name: Hosted Linux Preview
+    ${{ if eq(parameters.os, 'Windows_NT') }}:
+      name: Hosted VS2017
+  steps:
+  - checkout: self
+    clean: true
+  - ${{ parameters.beforeBuild }}
+  - ${{ if eq(parameters.os, 'Windows_NT') }}:
+    - script: .\build.cmd -ci ${{ parameters.buildArgs }}
+      displayName: Run build.cmd
+  - ${{ if ne(parameters.os, 'Windows_NT') }}:
+    - script: ./build.sh -ci ${{ parameters.buildArgs }}
+      displayName: Run build.sh
+  - task: PublishTestResults@2
+    displayName: Publish test results
+    condition: always()
```

Comment:
Consider `succeededOrFailed()` vs `always()`.  The only difference being that the former doesn't execute if the build was cancelled.

Reply:
In general, yes `succeededOrFailed` is probably good, but even if a build is cancelled, I want to see test results for test projects that finished before cancellation. For example, if i have to cancel a build that is gets hung while running tests...which happens a lot to Kestrel.